contracts/zapper-contract/SnowglobeZapperVector.sol

```diff
@@ -0,0 +1,140 @@
+        path[0] = swapToken;
+        path[1] = desiredToken;
+
+        _approveTokenIfNeeded(path[0], address(router));
+        IJoeRouter(router).swapExactTokensForTokens(
+            IERC20(swapToken).balanceOf(address(this)),
+            desiredTokenOutMin,
+            path,
+            address(this),
+            block.timestamp
+        );
+
+        _returnAssets(path);
+    }
+
+    function _swapAndStake(address snowglobe, uint256 tokenAmountOutMin, address tokenIn) public override {
+        (IGlobe vault, IUniPair pair) = _getVaultPair(snowglobe);
+
+        (uint256 reserveA, uint256 reserveB, ) = pair.getReserves();
+        require(reserveA > minimumAmount && reserveB > minimumAmount, "Liquidity pair reserves too low");
+
+        address token0 = pair.token0();         // first token in the pool
+        address token1 = pair.token1();         // second token in the pool 
+
+        bool isInputA = token0 == tokenIn;
+        require(isInputA || token1 == tokenIn, "Input token not present in liquidity pair");
+
+        address[] memory path = new address[](2);
+        path[0] = tokenIn;
+        path[1] = isInputA ? token1 : token0;
+
+        uint256 fullInvestment = IERC20(tokenIn).balanceOf(address(this));
+        uint256 swapAmountIn;
+        if (isInputA) {
+            swapAmountIn = _getSwapAmount(fullInvestment, reserveA, reserveB);
+        } else {
+            swapAmountIn = _getSwapAmount(fullInvestment, reserveB, reserveA);
+        }
+
+        _approveTokenIfNeeded(path[0], address(router));
+         uint256[] memory swappedAmounts = IJoeRouter(router)
+            .swapExactTokensForTokens(
+                swapAmountIn,
+                tokenAmountOutMin,
+                path,
+                address(this),
+                block.timestamp
+            );
+
+        _approveTokenIfNeeded(path[1], address(router));
+        (, , uint256 amountLiquidity) = IJoeRouter(router).addLiquidity(
+            path[0],
+            path[1],
+            fullInvestment.sub(swappedAmounts[0]),
+            swappedAmounts[1],
+            1,
+            1,
+            address(this),
+            block.timestamp
+        );
+
+        _approveTokenIfNeeded(address(pair), address(vault));
+        vault.deposit(amountLiquidity);
+
+        //taking receipt token and sending back to user
+        vault.safeTransfer(msg.sender, vault.balanceOf(address(this)));
+
+        //interact with gauge proxy to get gauge address
+        address gaugeAddress = IGaugeProxyV2(gaugeProxy).getGauge(snowglobe);
+
+        //deposit for into gauge
+        IGaugeV2(gaugeAddress).depositFor(vault.balanceOf(msg.sender), msg.sender);
+
+        _returnAssets(path);
+    }
+
+    function _getSwapAmount(uint256 investmentA, uint256 reserveA, uint256 reserveB) public view override returns (uint256 swapAmount) {
+        uint256 halfInvestment = investmentA.div(2);
+        uint256 nominator = IJoeRouter(router).getAmountOut(
```

Comment:
Why are we using the TraderJoe Router in the Vector Swap here?

Reply:
was gonna use kyber but it is more expensive in terms of gas fees, and traderjoe is the only one with liquidity for xptp-ptp swap